### ISSUE TITLE
Enable collecting metrics of Calico

### ISSUE THREAD

**Issue by invidian** [kinvolk/lokomotive#205]
Felix from Calico exposes metrics in Prometheus format, we should be collecting them via ServiceMonitor object with Prometheus Operator.

However, during cluster bootstrapping, ServiceMonitor CRD is not available, so it cannot be deployed then. We should figure out how to deploy it in a single step, so collecting metrics can be test with CI.

One and probably the simplest idea would be to have a separate component, which will only contain ServiceMonitor for Calico and must be installed **after** Prometheus Operator component. However, that solution doesn't seem very "integrated".

Another approach would be to add Calico ServiceMonitor object to Prometheus Operator helm chart, but I am not sure if such patch would be accepted upstream (as they would have to accept monitoring for all CNIs).

**Comment by invidian** [comment]
For the reference, following patch enables collection of the metrics:
```patch
diff --git a/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/daemonset.yaml b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/daemonset.yaml
index c5baa89d..cdf1bfbc 100644
--- a/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/daemonset.yaml
+++ b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/daemonset.yaml
@@ -65,6 +65,9 @@ spec:
       containers:
         - name: calico-node
           image: {{ .Values.calico.image }}
+          ports:
+          - containerPort: 9091
+            name: metrics
           env:
             # Use Kubernetes API as the backing datastore.
             - name: DATASTORE_TYPE
diff --git a/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/service.yaml b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/service.yaml
new file mode 100644
index 00000000..d19b1f29
--- /dev/null
+++ b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/service.yaml
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: calico-node-metrics
+  namespace: kube-system
+  labels:
+    k8s-app: calico-node
+spec:
+  ports:
+  - port: 9091
+    name: metrics
+  selector:
+    k8s-app: calico-node
+  type: ClusterIP
diff --git a/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/servicemonitor.yaml b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/servicemonitor.yaml
new file mode 100644
index 00000000..1b79b283
--- /dev/null
+++ b/assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/servicemonitor.yaml
@@ -0,0 +1,16 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: calico-node
+  namespace: kube-system
+  labels:
+    release: prometheus-operator
+spec:
+  endpoints:
+  - port: metrics
+  namespaceSelector:
+    matchNames:
+    - kube-system
+  selector:
+    matchLabels:
+      k8s-app: calico-node
```

**Comment by surajssd** [comment]
I wonder how is this test passing then :thinking:  https://github.com/kinvolk/lokomotive/blob/903ed9d00ffea55b291a413f42da8803fe590312/test/monitoring/control-plane_test.go#L59-L62

**Comment by invidian** [comment]
>I wonder how is this test passing then thinking, without above patch?

I am not sure. Will investigate. By looking at the manifests we have in the repo, I see now way those metrics could be collected, but maybe there is some magic happening, which I'm not aware of.

**Comment by invidian** [comment]
@surajssd do you know the status of it? I recall you had some port collision around it, right?

**Comment by surajssd** [comment]
The issue was this https://github.com/kinvolk/lokomotive/issues/298

But about calico metrics, I am not sure.

**Comment by kskewes** [comment]
Prometheus Operator now supports [`PodMonitor`](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#podmonitor) which negates the need for doing a `Service`.

We use them a lot and use it for Calico metrics, both `Typha` and `Felix` per https://docs.projectcalico.org/maintenance/monitor/monitor-component-metrics

For AWS we have to modify their calico.yaml to expose the metrics ports on both.
For AWS we have to manage the control plane components anyway so we bring the yaml's in house and sync with AWS upstream.

For Lokomotive I don't know what direction the project would like to take.
On one hand we could expose the ports if we wanted to enable Prometheus considering Calico is bundled in.
As a single sample, I will bring my own Prometheus and most components.

If Lokomotive enabled BYO CNI then perhaps exposing Calico metrics ports doesn't make sense and users could choose to self-manage Calico as well?

Would something like a Prometheus Operator Lokomotive Extensions helm chart work?
opt-in to have various ServiceMonitors/PodMonitors installed and that can be done after Prometheus Operator but in a seamless way?
Similar situation for MetalLB and others?

**Comment by invidian** [comment]
@kskewes thanks for looking into it. As far as I know, the port is already exposed (`ports` in pod spec are information only, ports will be exposed anyway if application is listening), but the tricky part here is installing the ServiceMonitor object.

I think _ideally_, we would have a way to customize an upstream `kube-prometheus-stack` to include a `ServiceMonitor` object for Calico there, without modifying it at the source level (we could use Helm's post-renderer functionality). And that would work fine for now.

But I think it's not a huge deal to just add an extra file to the chart for now, as we do it anyway for some charts. And extra file is easier to maintain than patch on existing files.

If in the future we decide to add (back) support for different CNIs, then we would have to pass the CNI information to the prometheus-operator component (#499) and with that, we would include appropriate `ServiceMonitor` and other objects.

TL;DR Let's add `Service` and `ServiceMonitor` for Calico to `prometheus-operator` component, e2e tests to ensure metrics are scraped and we can close this issue.
